### ISSUE TITLE
interactor.selectDate doesn't work in PhantomJS

### ISSUE THREAD

**Issue by marcoow** [edgycircle/ember-pikaday#16]
`interactor.selectDate` doesn't work in PhantomJS. While after calling the method `interactor.selectedMonth()` and `interactor.selectedYear()`, `interactor.selectedDay()` is undefined. I believe that has to do with the `mouseDown` event that's triggered for selecting the day - I guess PhantomJS doesn't support that somehow or at least not the way you're firing it. I'm running PhantomJS 2.0.0.

**Comment by stravid** [comment]
Thanks for the bug report! I will upgrade to the new PhantomJS myself and debug it.

**Comment by stravid** [comment]
Ok, I managed to install PhantomJS 2.0.0 and run the ember-pikaday tests itself. Strangely enough they all pass. I assume you are using ember-pikaday in a project and the projects acceptance test fails, correct? Can you tell me what version of Ember CLI you are using so I can setup a dummy project on my own and figure out what is happening? Thanks!

```
~/Code/ember-pikaday master*  17m 16s
❯ ember test
version: 0.1.15
Could not find watchman, falling back to NodeWatcher for file system events
Built project successfully. Stored in "/Users/david/Code/ember-pikaday/tmp/class-tests_dist-oh961SA0.tmp".
ok 1 PhantomJS 2.0 - JSHint - modules/ember-pikaday/components: modules/ember-pikaday/components/pikaday-input.js should pass jshint
ok 2 PhantomJS 2.0 - JSHint - modules/ember-pikaday/helpers: modules/ember-pikaday/helpers/pikaday.js should pass jshint
ok 3 PhantomJS 2.0 - JSHint - .: app.js should pass jshint
ok 4 PhantomJS 2.0 - JSHint - helpers: helpers/resolver.js should pass jshint
ok 5 PhantomJS 2.0 - JSHint - helpers: helpers/start-app.js should pass jshint
ok 6 PhantomJS 2.0 - JSHint - initializers: initializers/setup-pikaday-i18n.js should pass jshint
ok 7 PhantomJS 2.0 - JSHint - .: router.js should pass jshint
ok 8 PhantomJS 2.0 - JSHint - .: test-helper.js should pass jshint
ok 9 PhantomJS 2.0 - PikadayInputComponent: is an input tag
ok 10 PhantomJS 2.0 - PikadayInputComponent: the input tag has the readonly attribute if it has been set on the component
ok 11 PhantomJS 2.0 - PikadayInputComponent: clicking the input opens the pikaday dialog
ok 12 PhantomJS 2.0 - PikadayInputComponent: selecting a date should update the value attribute
ok 13 PhantomJS 2.0 - PikadayInputComponent: setting the value attribute should select the correct date
ok 14 PhantomJS 2.0 - PikadayInputComponent: DD.MM.YYYY should be the default format for the input
ok 15 PhantomJS 2.0 - PikadayInputComponent: format of the input is changeable
ok 16 PhantomJS 2.0 - PikadayInputComponent: yearRange of the input defaults to 10
ok 17 PhantomJS 2.0 - PikadayInputComponent: yearRange of the input can be set with a range
ok 18 PhantomJS 2.0 - PikadayInputComponent: yearRange of the input can be set with comma separated years
ok 19 PhantomJS 2.0 - PikadayInputComponent: yearRange of the input with comma separated years supports currentYear as max
ok 20 PhantomJS 2.0 - PikadayInputComponent: default i18n configuration of Pikaday can be changed
ok 21 PhantomJS 2.0 - PikadayInputComponent: if utc is set the date returned from pikaday should be in UTC format
ok 22 PhantomJS 2.0 - JSHint - unit/components: unit/components/pikaday-input-test.js should pass jshint

1..22
# tests 22
# pass  22
# fail  0
```

**Comment by stravid** [comment]
And while we are at it, what version of Pikaday do you use?

**Comment by marcoow** [comment]
we're using 

```json
…
"ember-cli": "^0.1.15",
"ember-pikaday": "dertoti/ember-pikaday#master",
…
```

**Comment by marcoow** [comment]
It looks like pikaday's [`mouseDown` event handler](https://github.com/dbushell/Pikaday/blob/f2c8d56fb8bd65cdc702d212b2df09a6742d728f/pikaday.js#L1017) is never called in our acceptance test. Maybe that has sth. to do with Ember registering its own `mouseDown` handler which isn't set up in ember-pikaday's tests as those don't start up an Ember app.

**Comment by stravid** [comment]
Thanks for the info, if you are running the not yet merged pull request the culprit is likely the newer version of Pikaday. I will start updating the Pikaday version used by ember-pikaday and look where it starts breaking so it can be fixed.

**Comment by marcoow** [comment]
No, the same problem exists with `"ember-pikaday": "0.6.0"` as well.

**Comment by stravid** [comment]
Ok, thanks again for the info.

**Comment by stravid** [comment]
Does #18 solve your problem?

Because I created a dummy application and can reproduce the error, it seems that the Pikaday DOM element with the days is not yet in the DOM when `triggerNativeEvent($('td[data-day="' + day + '"] button')[0], 'mousedown');` is called. Therefore the jQuery selector returns `undefined` and it fails.

I think that's the problem, I'm currently failing to find a method to wait until the DOM is populated by Pikaday to test my theory.

**Comment by marcoow** [comment]
No, like I wrote in #18 the problem is that Pikaday master binds to `'ontouchend'` if that's available (which is the case in PhantomJS) while ember-pikaday's `selectDate` method always triggers a `'mousedown'` event so that Pikaday's event handler is never invoked.

The point is that ember-pikaday doesn't currently work with Pikaday master.

**Comment by stravid** [comment]
I'm not sure how to interpret the "No, ..." :-)
Does #18 solve your problem? If triggering the correct event is solving your situation then I can merge it in and your tests work correctly.

Nonetheless there seems to be another problem with the test helper.

**Comment by marcoow** [comment]
Yes, #18 solves my problem but no, I don't think your above explanation for the error is correct.

**Comment by stravid** [comment]
Very good, should I cut a new version after merging or is using `ember-pikaday#master` okay for the time being? I intend to release a version anyways but want to fix some other things first...

It seems there are two problems, the one you discovered, and one I discovered in my test application while trying to fix your problem.

**Comment by marcoow** [comment]
I'm using [this branch](https://github.com/simplabs/ember-pikaday/tree/termify) now anyway. I don't know whether you'd want to publish a release of `ember-pikaday` that depends on `pikaday#master`… - probably better to wait for the next version of pikaday before merging #18 .

**Comment by kristianmandrup** [comment]
I did a lot of work in my branch to make it integrate better with Ember:

https://github.com/kristianmandrup/Pikaday

**Comment by stravid** [comment]
Alright, then I'm not in a hurry. Thanks for digging in and finding the solution @marcoow!

@kristianmandrup Until your branch gets merged into Pikaday this doesn't help us much here.

**Comment by kristianmandrup** [comment]
Yeah, it’s a pity my additions were never merged. A lot of good stuff there, especially for dynamic data binding, customization and more control

**Comment by Robdel12** [comment]
Any ideas if this will be fixed? Phantom tests fail for me.

**Comment by jrjohnson** [comment]
My tests started failing today when I did a bower install with Pikaday 1.3.3 (released two days ago).  Locking the version to 1.3.2 in bower caused everything to pass again.

**Comment by stravid** [comment]
So Pikaday finally tagged the new version, I was checking myself every other week but thanks for notifying me. A new version with this issue fixed will be released on Monday!

**Comment by stravid** [comment]
Resolved with release `0.8.0`.
